bigquery/src/main/java/org/apache/iceberg/gcp/bigquery/BigQueryMetastoreCatalog.java

```diff
@@ -0,0 +1,400 @@
+  public static final String TESTING_ENABLED = "testing-enabled";
+
+  public static final String HIVE_METASTORE_WAREHOUSE_DIR = "hive.metastore.warehouse.dir";
+
+  private static final Logger LOG = LoggerFactory.getLogger(BigQueryMetastoreCatalog.class);
+
+  private static final String DEFAULT_GCP_LOCATION = "us";
+
+  private String catalogName;
+  private Map<String, String> catalogProperties;
+  private FileIO fileIO;
+  private Configuration conf;
+  private String projectId;
+  private String location;
+  private BigQueryMetastoreClient client;
+  private boolean filterUnsupportedTables;
+
+  // Must have a no-arg constructor to be dynamically loaded
+  // initialize(String name, Map<String, String> properties) will be called to complete
+  // initialization
+  public BigQueryMetastoreCatalog() {}
+
+  @Override
+  public void initialize(String inputName, Map<String, String> properties) {
+    Preconditions.checkArgument(
+        properties.containsKey(PROJECT_ID), "GCP project must be specified");
+
+    projectId = properties.get(PROJECT_ID);
+    location = properties.getOrDefault(GCP_LOCATION, DEFAULT_GCP_LOCATION);
+    boolean testingEnabled =
+        Boolean.parseBoolean(properties.getOrDefault(TESTING_ENABLED, "false"));
+
+    BigQueryOptions options =
+        BigQueryOptions.newBuilder()
+            .setProjectId(projectId)
+            .setLocation(location)
+            .setRetrySettings(ServiceOptions.getDefaultRetrySettings())
+            .build();
+
+    try {
+      if (testingEnabled) {
+        client = new FakeBigQueryMetastoreClient(options);
+
+      } else {
+        client = new BigQueryMetastoreClientImpl(options);
+      }
+    } catch (IOException e) {
+      throw new ServiceFailureException(e, "Creating BigQuery client failed");
+    } catch (GeneralSecurityException e) {
+      throw new ValidationException(e, "Creating BigQuery client failed due to a security issue");
+    }
+    initialize(inputName, properties, projectId, location, client);
+  }
+
+  @VisibleForTesting
+  void initialize(
+      String inputName,
+      Map<String, String> properties,
+      String initialProjectId,
+      String initialLocation,
+      BigQueryMetastoreClient bigQueryMetaStoreClient) {
+    Preconditions.checkArgument(bigQueryMetaStoreClient != null, "BigQuery client can not be null");
+    this.catalogName = inputName;
+    this.catalogProperties = ImmutableMap.copyOf(properties);
+    this.projectId = initialProjectId;
+    this.location = initialLocation;
+    this.client = bigQueryMetaStoreClient;
+
+    if (this.conf == null) {
+      LOG.warn("No configuration was set, using the default environment Configuration");
+      this.conf = new Configuration();
+    }
+
+    LOG.info("Using BigQuery Metastore Iceberg Catalog: {}", inputName);
+
+    if (properties.containsKey(CatalogProperties.WAREHOUSE_LOCATION)) {
+      // Iceberg always removes trailing slash to avoid paths like "<folder>//data" in file systems
```

Comment:
I would probably just remove this comment

Reply:
Done